### PR TITLE
Add support for raw_input_contents and raw_output_contents in gRPC

### LINKED ISSUE #48: Add support for raw_input_contents field
Add support for top-level `raw_input_contents` in gRPC spec. This field was initially introduced to work-around some performance issues within gRPC.

You can find more details on these issues: 

- https://github.com/triton-inference-server/server/issues/1821
- https://github.com/kubeflow/kfserving/pull/998

### PR DESCRIPTION
**Pull request by adriangonz**
Fixes #48

## Changelog
- Update proto to include `raw_input_contents` and `raw_output_contents`
- Include new `mlserver.raw` module with some utilities to unpack / pack data into raw format
- Update gRPC converters module to take into account the new raw fields

### CODE REVIEW

**Comment by cliveseldon** [comment]
Looks good. One question when you pack back into raw contents to you remove the explicit data contents as with Triton you get one set but never both? Just wasn't clear on first glance.

File: mlserver/grpc/converters.py [hunk 1]
```diff
@@ -310,6 +324,10 @@ def from_types(cls, type_object: types.InferenceResponse) -> pb.ModelInferRespon
             ],
         )
 
+        if use_raw:
+            # If using raw outputs, ensure it's set on the final object
+            model_infer_response.raw_output_contents.extend(raw)
```
Comment (sakoush) [inline mlserver/grpc/converters.py:329]: The assumption here is that raw contents is always the last field in the grpc message, which is fine with the current spec. I just wonder is the spec guarantees this is always the case? it makes sense to be always the case but just checking.

File: mlserver/grpc/converters.py [hunk 2]
```diff
@@ -135,6 +136,12 @@ def to_types(cls, pb_object: pb.ModelInferRequest) -> types.InferenceRequest:
                 for out in pb_object.outputs
             ]
 
+        if pb_object.raw_input_contents:
+            # Unpack and inject raw contents into `data` fields if present
+            inject_raw(
+                inference_request.inputs, pb_object.raw_input_contents  # type: ignore
```
Comment (sakoush) [inline mlserver/grpc/converters.py:142]: it wasnt very clear that grpc takes care of returning a list of bytes in `raw_input_contents` one for each input request. I thought initially this is one big stream of bytes but maybe it is only me.

anyway perhaps add a note to make it clearer?

**Review by sakoush — APPROVED** [review]
Very neat! I have (general ) questions:
- how safe is it if we are using raw_contents when we are chaining services from different architecture, os (e.g. unix and windows) and/or python versions? is `struct` taking care of all of that? is it better or worse with triton?
- is it possible to combine codecs and raw_contents? i.e. how pandas dataframes are being sent if raw_contents is enabled? as numpy arrays and there is a further transform to convert to pandas df on the receiver side perhaps?
- a minor question whether the spec guarantees that raw_contents is the last field in the grpc message?

File: tests/test_raw.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+def test_unpack_bytes():
+    raw = b"\x03\x00\x00\x00one\x03\x00\x00\x00two\x05\x00\x00\x00three"
+    expected = [b"one", b"two", b"three"]
+
+    unpacked = unpack_bytes(raw)
+    assert unpacked == expected
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (adriangonz) [reply tests/test_raw.py:26]: In the test case, these are not strings but bytes arrays already.

File: mlserver/raw.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return reduce(mul, shape, 1)
+
+
+def _tensor_format(elem: InputOrOutput) -> str:
+    size = _tensor_length(elem.shape)
+    ctype = _DatatypeToCtype[elem.datatype]
+    return f"{size}{ctype}"
+
+
+def unpack_bytes(raw: bytes) -> List[bytes]:
```
[the remark this answers is not in the document]
Comment (adriangonz) [reply mlserver/raw.py:39]: Maybe? I think initially these were on a separate module, and then I moved them all here. 

I could make them private though :+1:

File: mlserver/grpc/converters.py (hunk 1 above)
Comment (sakoush) [reply mlserver/grpc/converters.py:329]: oh yes I just misread the code. it is fine.

File: mlserver/grpc/converters.py (hunk 2 above)
Comment (sakoush) [reply mlserver/grpc/converters.py:142]: ok  fine if there is a note somewhere else.

**Comment by adriangonz** [comment]
Thanks for your review @sakoush !

To answer the points on the main comment:

    - The implementation is very similar to Triton's. Bytes are actually encoded on the same way. The only difference is with tensors, where we use `struct` and they use the `tobytes()` and `frombuffer()` helpers from `numpy`. Both should be equivalent though. Architecture-wise, the `BYTES` encoding sets explicitly the endianness to ensure there're no issues. Perhaps we could investigate what `numpy` does for tensors to ensure we do the same thing? 
    -  Yes, it should. The `raw -> InferenceRequest` conversion occurs before any codecs are applied. So after that, the normal MLServer flow should kick off. Regarding how to write the request, same principle applies. You could use the `PandasCodec.encode_request()` helper and then call the `mlserver.raw.extract_raw` method to convert the normal data to raw (or you could do it using `tritonclient`, or manually). 
    - Not sure what you mean about this one. On both Triton and MLServer, the `raw_input_contents` / `raw_output_contents` fields are set on the same position of their respective protobuf objects (i.e. `7` and `6`).

**Comment by sakoush** [comment]
> * `tobytes()` and `frombuffer()` helpers from `numpy`.

the ultimate test is to have mlserver -> triton inference pipeline with raw_contents perhaps which we can do with your change :)

> fields are set on the same position of their respective protobuf objects (i.e. 7 and 6).

I am not saying they are different I am just trying to think a bit in the future and consider if there are any further optimisations that we could do, especially in SCv2 pipelines if we know that the raw_contents happen to be the last thing. I dont have anything concrete yet thought that we could explore.

**Pull request merged**